File: MinHeap.java
public class MinHeap {
    private HuffmanNode[] heap;
    private int size;
    private int capacity;

    public MinHeap(int capacity) {
        this.capacity = capacity;
        this.heap = new HuffmanNode[capacity];
        this.size = 0;
    }

    private void swap(int i, int j) {
        HuffmanNode temp = heap[i];
        heap[i] = heap[j];
        heap[j] = temp;
    }

    private void heapify(int i) {
        int smallest = i;
        int left = 2 * i + 1;
        int right = 2 * i + 2;

        if (left < size && heap[left].frequency < heap[smallest].frequency)
            smallest = left;
        if (right < size && heap[right].frequency < heap[smallest].frequency)
            smallest = right;
        if (smallest != i) {
            swap(i, smallest);
            heapify(smallest);
        }
    }

    public void insert(HuffmanNode node) {
        if (size == capacity) {
            HuffmanNode[] newHeap = new HuffmanNode[capacity * 2];
            System.arraycopy(heap, 0, newHeap, 0, capacity);
            heap = newHeap;
            capacity *= 2;
        }
        int i = size;
        heap[size++] = node;

        while (i != 0) {
            int parent = (i - 1) / 2;
            if (heap[parent].frequency > heap[i].frequency) {
                swap(parent, i);
                i = parent;
            } else break;
        }
    }

    public HuffmanNode extractMin() {
        if (size == 0) return null;
        HuffmanNode root = heap[0];
        heap[0] = heap[size - 1];
        heap[size - 1] = null;
        size--;
        heapify(0);
        return root;
    }

    public static MinHeap buildHeap(HuffmanNode[] nodes, int n) {
        MinHeap mh = new MinHeap(Math.max(1, n));
        mh.size = n;
        mh.capacity = Math.max(1, n);
        mh.heap = new HuffmanNode[mh.capacity];
        System.arraycopy(nodes, 0, mh.heap, 0, n);

        for (int i = (mh.size / 2) - 1; i >= 0; i--)
            mh.heapify(i);
        return mh;
    }

    public int getSize() { return size; }
}

File: minheap.c
#include <stdlib.h>
#include "minheap.h"

static void swap_nodes(HuffmanNode **a, HuffmanNode **b) {
    HuffmanNode *t = *a; *a = *b; *b = t;
}

MinHeap *create_minheap(int capacity) {
    MinHeap *h = (MinHeap*) malloc(sizeof(MinHeap));
    if (!h) return NULL;
    h->capacity = capacity > 0 ? capacity : 1;
    h->size = 0;
    h->array = (HuffmanNode**) malloc(sizeof(HuffmanNode*) * h->capacity);
    if (!h->array) { free(h); return NULL; }
    return h;
}

void minheap_free(MinHeap *h) {
    if (!h) return;
    free(h->array);
    free(h);
}

static void heapify_down(MinHeap *h, int idx) {
    int smallest = idx;
    int l = 2*idx + 1;
    int r = 2*idx + 2;
    if (l < h->size && h->array[l]->freq < h->array[smallest]->freq) smallest = l;
    if (r < h->size && h->array[r]->freq < h->array[smallest]->freq) smallest = r;
    if (smallest != idx) {
        swap_nodes(&h->array[smallest], &h->array[idx]);
        heapify_down(h, smallest);
    }
}

static void heapify_up(MinHeap *h, int idx) {
    while (idx && h->array[(idx-1)/2]->freq > h->array[idx]->freq) {
        swap_nodes(&h->array[idx], &h->array[(idx-1)/2]);
        idx = (idx-1)/2;
    }
}

void minheap_insert(MinHeap *h, HuffmanNode *node) {
    if (h->size == h->capacity) {
        int newcap = h->capacity * 2;
        HuffmanNode **tmp = (HuffmanNode**) realloc(h->array, sizeof(HuffmanNode*) * newcap);
        if (!tmp) return;
        h->array = tmp;
        h->capacity = newcap;
    }
    h->array[h->size] = node;
    heapify_up(h, h->size);
    h->size++;
}

HuffmanNode *minheap_extract_min(MinHeap *h) {
    if (h->size == 0) return NULL;
    HuffmanNode *root = h->array[0];
    h->array[0] = h->array[h->size - 1];
    h->size--;
    heapify_down(h, 0);
    return root;
}

void build_minheap_from_table(MinHeap *h, unsigned long freq_table[]) {
    for (int i = 0; i < ALPHABET; ++i) {
        if (freq_table[i] > 0) {
            HuffmanNode *n = (HuffmanNode*) malloc(sizeof(HuffmanNode));
            n->ch = (unsigned char)i;
            n->freq = freq_table[i];
            n->left = n->right = NULL;
            minheap_insert(h, n);
        }
    }
}